src/main/scala/cloud/drdrdr/utils/AWSCredentials.scala

```diff
@@ -6,165 +6,416 @@ package cloud.drdrdr.utils
+import akka.stream.scaladsl.Sink
+import akka.util.ByteString
 import spray.json._
+import org.slf4j.LoggerFactory
+import com.typesafe.scalalogging.Logger
 import scala.concurrent.duration._
-import scala.concurrent.{Promise, Await, ExecutionContext, Future}
-import scala.util.{Success, Failure}
+import scala.concurrent.{ExecutionContext, Future, Await}
+import scala.io.Source
+import scala.util.control
+
+
+trait AWSCredentials {
+
+  case class AWSCredentialSource(creds: Future[AWSPermissions]) (implicit ec: ExecutionContext, s:ActorSystem, m: ActorMaterializer) {
+    private var credentials: Future[AWSPermissions] = creds
+
+    /**
+     * Updates the credentials if necessary, then returns the credentials
+     * @return AWSPermissions with most recent credentials
+     */
+    def getCredentials: Future[AWSPermissions] = {
+      credentials flatMap {
+        case perm:AWSPermissions =>
+          if (perm.expiration.isEmpty) {
+              credentials
+          } else {
+              checkExpire flatMap {
+                case bol =>
+                  if (bol) {
+                    credentials = updateCredentials()
+                  }
+                  credentials
+              }
+          }
+      }
+    }
+
+    //checks if current credentials have expired
+    private def checkExpire()(implicit ec: ExecutionContext, s: ActorSystem, m: ActorMaterializer): Future[Boolean] = {
+      credentials map {
+        case creds: AWSPermissions =>
+          if (creds.expiration.isEmpty)
+            false
+          else
+            creds.expiration <= getUTCTime()
+      }
+    }
+//
+    //updates the credentials on an EC2 instance if necessary
+    @throws(classOf[Exception])
+    private def updateCredentials(): Future[AWSPermissions] = {
+      getAmazonEC2Credentials() map {
+        case Some(cred) => cred
+        case None => throw new Exception("Unable to update credentials")
+      }
+    }
+  }
 
+  /**
+   *
+   * @param keyId aws accessKeyId
+   * @param secretKey aws accessSecretKey
+   * @param tok aws token associated with credentials
+   * @param exp aws expiration of credentials
+   */
+  class AWSPermissions(keyId: String, secretKey: String, tok: String = "", exp: String = "") {
+    private val accessKey = keyId
+    private val secretAccess = secretKey
+    private val t = tok
+    private val e = exp
 
-object AWSCredentials {
+    /**
+     * accessKeyId accessor method
+     * @return accessKeyId
+     */
+    def accessKeyId: String = {
+      this.accessKey
+    }
 
-  implicit val system: ActorSystem = ActorSystem()
-  implicit val ec: ExecutionContext = system.dispatcher
-  implicit val materializer: ActorMaterializer = ActorMaterializer()
+    /**
+     * secretAccessKey accessor method
+     * @return secretAccessKey
+     */
+    def secretAccessKey: String = {
+      this.secretAccess
+    }
 
-  case class AWSPermissions(accessKeyId: String, secretAccessKey: String, token: String = "")
```

Comment:
why did you choose to go with a class instead of a case class here? There's only accessor methods in here.